src/listpack.c

```diff
@@ -787,25 +885,79 @@ unsigned char *lpInsert(unsigned char *lp, unsigned char *ele, uint32_t size, un
     return lp;
 }
 
+/* This is just a wrapper for lpInsert() to insert a string, which is
+ * Possible a integer encoded strings. */
+unsigned char *lpInsertString(unsigned char *lp, unsigned char *ele, uint32_t size, unsigned char *p, int where, unsigned char **newp) {
+    int enctype;
+    uint64_t enclen; /* The length of the encoded element. */
+    unsigned char intenc[LP_MAX_INT_ENCODING_LEN];
+
+    /* Calling lpEncodeGetType() results into the encoded version of the
+     * element to be stored into 'intenc' in case it is representable as
+     * an integer: in that case, the function returns LP_ENCODING_INT.
+     * Otherwise if LP_ENCODING_STR is returned, we'll have to call
+     * lpEncodeString() to actually write the encoded string on place later.
+     *
+     * Whatever the returned encoding is, 'enclen' is populated with the
+     * length of the encoded element. */
+    if (ele) {
+        enctype = lpEncodeGetType(ele,size,intenc,&enclen);
+    } else {
+        enctype = -1;
+        enclen = 0;
+    }
+
+    return lpInsert(lp, enctype, enclen, ele, size, intenc, p, where, newp);
+}
+
+/* This is just a wrapper for lpInsert() to directly use a 64 bit integer
+ * instead of a string. */
+unsigned char *lpInsertInteger(unsigned char *lp, long long lval, unsigned char *p, int where, unsigned char **newp) {
+    uint64_t enclen; /* The length of the encoded element. */
+    unsigned char intenc[LP_MAX_INT_ENCODING_LEN];
+
+    lpEncodeIntegerGetType(lval,intenc,&enclen);
+    return lpInsert(lp, LP_ENCODING_INT, enclen, NULL, 0, intenc, p, where, newp);
+}
+
+/* Append the specified element 's' of length 'slen' at the head of the listpack. */
+unsigned char *lpPrepend(unsigned char *lp, unsigned char *s, uint32_t slen) {
+    unsigned char *p = lpFirst(lp);
+    if (!p) return lpAppend(lp, s, slen);
+    return lpInsertString(lp, s, slen, p, LP_BEFORE, NULL);
+}
+
 /* Append the specified element 'ele' of length 'len' at the end of the
  * listpack. It is implemented in terms of lpInsert(), so the return value is
  * the same as lpInsert(). */
 unsigned char *lpAppend(unsigned char *lp, unsigned char *ele, uint32_t size) {
     uint64_t listpack_bytes = lpGetTotalBytes(lp);
     unsigned char *eofptr = lp + listpack_bytes - 1;
-    return lpInsert(lp,ele,size,eofptr,LP_BEFORE,NULL);
+    return lpInsertString(lp,ele,size,eofptr,LP_BEFORE,NULL);
+}
+
+/* Append the specified integer element 'lval' at the end of the listpack. */
+unsigned char *lpAppendInteger(unsigned char *lp, int64_t lval) {
+    uint64_t listpack_bytes = lpGetTotalBytes(lp);
+    unsigned char *eofptr = lp + listpack_bytes - 1;
+    return lpInsertInteger(lp,lval,eofptr,LP_BEFORE,NULL);
```

Comment:
shall we also add lpPrependInteger, and lpReplaceInteger right away, or wait for someone that needs it?

actually i see there's already one in t_stream.c, let's delete it from there and both of these additional interfaces here.

Reply:
`lpPrependInteger` and `lpReplaceInteger` have been added.